docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1055 @@
+the same generation validation requirement.
+
+### Pageserver attach/startup changes
+
+#### Attachment
+
+Calls to `/v1/tenant/{tenant_id}/attach` are augmented with an additional
+`generation` field in the body.
+
+The pageserver does not persist this: a generation is only good for the lifetime
+of a process.
+
+#### Re-attachment on startup
+
+On startup, the pageserver will call out to an new control plane `/re-attach`
+API (see [Generation API](#generation-api)). This returns a list of
+tenants that should be attached to the pageserver, and their generation numbers, which
+the control plane will increment before returning.
+
+The pageserver should still scan its local disk on startup, but should _delete_
+any local content for tenants not indicated in the `/re-attach` response: their
+absence is an implicit detach operation.
+
+#### Reconciliation
+
+On attachment, the `Tenant` may have some local disk content from previous
+activity involving this tenant (e.g. it used to be attached here), and this
+needs reconciling with remote storage. Note that "reconcile" in this context is
+used generically, and does not refer specifically to the logic currently
+in the `reconcile_with_remote` function.
+
+Because pageservers have access to the WAL, they don't strictly have to
+use the latest remote storage contents at all: if the WAL history still contains
+the data required to bring the pageserver's local state up to date with
+the tip of the WAL, the pageserver could recover from the WAL and then
+overwrite whatever is in remote storage.
+
+However, it is preferable to recover from remote storage rather than
+to replay WAL, for the following reasons:
+
+- Simplicity: recovering from remote storage gives identical behavior
+  whether attaching a tenant for the first time or re-attaching later.
+- Efficiency: ingesting the WAL has significant cost beyond the I/O
+  of reading it (compaction, image layer generation). We should avoid
+  doing this work twice if there is already remote content available.
+- Reduced load on safekeepers, which are a less scalable resource than S3.
+
+#### Finding the remote indices for timelines
+
+Because index files are now suffixed with generation numbers, the pageserver
+cannot always GET the remote index in one request, because it can't always
+know a-priori what the latest remote index is.
+
+Typically, the most recent generation to write an index would be our own
+generation minus 1. However, this might not be the case: the previous
+node might have started and acquired a generation number, and then crashed
+before writing out a remote index.
+
+In the general case and as a fallback, the pageserver may list all the `index_part.json`
+files for a timeline, sort them by generation, and pick the highest that is `<=`
+its current generation for this attachment. The tenant should never load an index
+with an attachment generation _newer_ than its own: tenants
```

Comment:
Don't mention migration here, which is an implicit forward reference to your next RFC.

Also, is my `Note` paragraph added in this suggestion correct? You don't have to add it to the doc, but, it wasn't totally obvious to me so I thought I'd write it down and check with you.

```suggestion
with an attachment generation _newer_ than its own.
These two rules combined ensure that objects written by later generations are never visible to earlier generations.

Note that if a given attachment picks an index part from an earlier generation (say n-2), but crashes & restarts before it writes its own generation's index part, next time it tries to pick an index part there may be an index part from generation n-1.
It would pick the n-1 index part in that case, because it's sorted higher than the previous one from generation n-2.
So, above rules guarantee no determinism in selecting the index part.
```

Reply:
It's kind of necessary to mention migration to motivate why we want tenants to work smoothly when running with an old generation: otherwise the reader might wonder why we let a stale generation'd attachment work at all.